File: torchsl/functional/multiview/mvcca.py
import torch
import torchsl

from .multiview_helpers import *

__all__ = ['mvcca']


# ===========================================================
# Multiview Canonical Correlation Analysis
# ===========================================================
def mvcca(Xs, C=True):
    if C:
        return torchsl.ops.mvcca(Xs)

    options = dict(dtype=Xs[0].dtype, device=Xs[0].device)
    num_views = len(Xs)
    num_components = Xs[0].size(0)
    num_samples = num_views * num_components
    dims = dimensions(Xs)

    I = torch.eye(num_components, **options)
    B = torch.ones(num_components, num_components, **options) / num_samples

    Sw = multiview_covariance_matrix(dims,
                                     lambda j, r: Xs[j].t().mm(I - B).mm(Xs[r]) if j == r
                                     else 0,
                                     options)
    Sb = multiview_covariance_matrix(dims,
                                     lambda j, r: Xs[j].t().mm(I - B).mm(Xs[r]) if j != r
                                     else 0,
                                     options)

    return Sw, Sb

File: torchsl/csrc/ops/multiview_subspace_learning/mvcca.cpp
#include <vector>
#include <tuple>
#include <torch/script.h>
#include "multiview_helpers.h"

namespace torchsl {
std::tuple<torch::Tensor, torch::Tensor> mvcca(
    const std::vector<torch::Tensor>& Xs
    ) {
    const auto options = Xs[0].options().requires_grad(false);
    const int num_views = Xs.size();
    const int num_components = Xs[0].size(0);
    const int num_samples = num_views * num_components;
    const auto dims = torchsl::dimensions(Xs);

    auto I = torch::eye(num_components, options);
    auto B = torch::ones({ num_components, num_components }, options).div_(num_samples);

    auto Sw = torchsl::multiview_covariance_matrix(dims, [&Xs, &I, &B] (int j, int r) {
        return (j == r) ? Xs[j].t().mm(I - B).mm(Xs[r]) : torch::tensor(0);
    }, options);
    auto Sb = torchsl::multiview_covariance_matrix(dims, [&Xs, &I, &B] (int j, int r) {
        return (j != r) ? Xs[j].t().mm(I - B).mm(Xs[r]) : torch::tensor(0);
    }, options);

    return std::make_tuple(Sw, Sb);
}
}
